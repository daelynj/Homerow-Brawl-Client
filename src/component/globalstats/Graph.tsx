import * as React from "react";
import { useState, useEffect } from "react";
import { Line } from "react-chartjs-2";

interface Props {
  gamesPlayed: any;
}

export const Graph = (props: Props) => {
  const [data, setData] = useState<any>({});

  useEffect(() => {
    let summedGameStats = sumGameStats();
    let labels = [];

    for (var index = 0; index < props.gamesPlayed.length; index++) {
      labels.push("");
    }

    const data = {
      labels: labels,
      datasets: [
        {
          label: "Average WPM",
          fill: false,
          lineTension: 0.1,
          backgroundColor: "rgba(75,192,192,0.4)",
          borderColor: "rgba(75,192,192,1)",
          borderCapStyle: "butt",
          borderDash: [],
          borderDashOffset: 0.0,
          borderJoinStyle: "miter",
          pointBorderColor: "rgba(0,0,0,1)",
          pointBackgroundColor: "#fff",
          pointBorderWidth: 3,
          pointHoverRadius: 5,
          pointHoverBackgroundColor: "rgba(75,192,192,1)",
          pointHoverBorderColor: "rgba(220,220,220,1)",
          pointHoverBorderWidth: 2,
          pointRadius: 1,
          pointHitRadius: 10,
          data: summedGameStats[0]
        },
        {
          label: "Average Mistakes",
          fill: false,
          lineTension: 0.1,
          backgroundColor: "rgba(255,165,0,0.4)",
          borderColor: "rgba(255,165,0,1)",
          borderCapStyle: "butt",
          borderDash: [],
          borderDashOffset: 0.0,
          borderJoinStyle: "miter",
          pointBorderColor: "rgba(0,0,0,1)",
          pointBackgroundColor: "#fff",
          pointBorderWidth: 3,
          pointHoverRadius: 5,
          pointHoverBackgroundColor: "rgba(255,165,0,1)",
          pointHoverBorderColor: "rgba(220,220,220,1)",
          pointHoverBorderWidth: 2,
          pointRadius: 1,
          pointHitRadius: 10,
          data: summedGameStats[1]
        },
        {
          label: "Average Accuracy",
          fill: false,
          lineTension: 0.1,
          backgroundColor: "rgba(255,0,0,0.4)",
          borderColor: "rgba(255,0,0,1)",
          borderCapStyle: "butt",
          borderDash: [],
          borderDashOffset: 0.0,
          borderJoinStyle: "miter",
          pointBorderColor: "rgba(0,0,0,1)",
          pointBackgroundColor: "#fff",
          pointBorderWidth: 3,
          pointHoverRadius: 5,
          pointHoverBackgroundColor: "rgba(255,0,0,1)",
          pointHoverBorderColor: "rgba(220,220,220,1)",
          pointHoverBorderWidth: 2,
          pointRadius: 1,
          pointHitRadius: 10,
          data: summedGameStats[2]
        }
      ]
    };

    setData(data);
  }, []);

  const sumGameStats = () => {
    let wpms: Array<number> = [];
    let mistakes: Array<number> = [];
    let accuracies: Array<number> = [];
    let WPMDataPoints: Array<number> = [];
    let mistakesDataPoints: Array<number> = [];
    let accuraciesDataPoints: Array<number> = [];

    for (var index = 0; index < props.gamesPlayed.length; index++) {
      wpms[index] = props.gamesPlayed[index].wpm;
      mistakes[index] = props.gamesPlayed[index].mistakes;
      accuracies[index] = props.gamesPlayed[index].accuracy;

      let sumWPM =
        wpms.reduce((a: number, b: number) => a + b, 0) / (index + 1);
      let sumMistakes =
        mistakes.reduce((a: number, b: number) => a + b, 0) / (index + 1);
      let sumAccuracies =
        accuracies.reduce((a: number, b: number) => a + b, 0) / (index + 1);

      WPMDataPoints[index] = Math.round(sumWPM);
      mistakesDataPoints[index] = Math.round(sumMistakes);
      accuraciesDataPoints[index] = Math.round(sumAccuracies);
    }

    return [WPMDataPoints, mistakesDataPoints, accuraciesDataPoints];
  };

  return (
    <>
      <Line data={data} />
    </>
  );
};
