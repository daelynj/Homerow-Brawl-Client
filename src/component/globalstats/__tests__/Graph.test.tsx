import * as React from "react";
import { Graph } from "../Graph";
import renderer from "react-test-renderer";

jest.mock("react-chartjs-2", () => ({
  Line: () => "Graph"
}));

describe("Graph", () => {
  let gamesPlayed = [
    { wpm: 80, mistakes: 3, accuracy: 100 },
    { wpm: 111, mistakes: 1, accuracy: 75 },
    { wpm: 96, mistakes: 7, accuracy: 83 },
    { wpm: 120, mistakes: 14, accuracy: 100 }
  ];

  it("renders a graph", () => {
    const tree = renderer.create(<Graph gamesPlayed={gamesPlayed} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
