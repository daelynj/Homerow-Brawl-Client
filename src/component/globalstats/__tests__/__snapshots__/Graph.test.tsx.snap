// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph renders a graph 1`] = `"Graph"`;
